### ISSUE TITLE
Update webpacksetup for Classic theme to accomodate for module specific changes

### ISSUE THREAD

**Issue by dennispw** [PrestaShop/PrestaShop#10478]
One of the core features (might even be a core principle?) in Prestashop is the ability to disable modules, override their assets on a theme level and more.

This is done easily by adding the correct structure in the theme `modules` folder which is great.

However, with the addition of the webpack workflow in `1.7` it would be extremely helpful if we could do the same thing with webpack.

**Right now** we have to manually add assets in the themes `modules` folder and edit them the "old fashioned" way:
```
/classic
  /modules
    /ps_imageslider
      [.css, .js, .tpl, etc]
```
It absolutely works and in comparison to previous versions of Prestashop it's a massive improvement to have it all in one place.

It would however be great if we somehow could automate it with webpack so we can do any module overrides using the already defined webpack workflow.

**As an example** if we we're to do this:
```
/classic
  /_dev
    /css
      /modules
        /ps_imageslider
          /css/
            homeslider.scss
```
It would be great if that somehow could be compiled into
```
/classic
  /modules
    /ps_imageslider
      /css
        homeslider.css
```
Of course, the same thing would be applied with `.js` files.

This means that we could much more quickly create proper overrides in the same "environment".

Automatically this would also mean that we don't loose the ability to _exclude_ a modules assets when making hook exceptions, uninstalling or deactivating modules.

Right now I can already see tendencies where _**module**_ assets are already making their way into the compiled `theme.css` which means that even if we uninstall and remove a module completely... It's assets will still be loaded because they are compiled in to the actual `theme.css`.

Just look at `ps_featuredproducts`. It has ties to the main `theme.css` with the `... .featured-products .product-miniature ...` CSS rules. Meaning the Classic theme would still be loading "unused assets" even if a certain shop has chosen _not_ to use that particular module and decided to uninstall it completely.

**To summarize**
Please think about using webpack to compile module assets so the main theme assets don't get to bloated and we don't loose the actual modular nature of Prestashop.

**Comment by khouloudbelguith** [comment]
@PrestaShop/prestashop-core-developers, @eternoendless  what do you think about his suggestion?  what do you think about his suggestion?

**Comment by eternoendless** [comment]
@CaptainYouz what do you think?

**Comment by eternoendless** [comment]
Hi @dennispw 

I discussed the issue with the team. We think that even though this _can_ be done using Webpack (it's basically about moving files around), we don't want to encourage people overriding module assets. 

File overriding means _replacing_ the module's assets with the theme's. This can quickly become an issue as any update made on the module will have to be ported back to the theme's copy, or it will stop working. Such a system doesn't scale, and even more so since unlike modules, PrestaShop doesn't currently have an update system for themes.

If you need to customize a part of the module's looks, it's best to add new styles which cascade over the module's. After all, that's what CSS was designed for 😄 

Of course, as you said, this means that your theme may load useless code if those modules aren't activated. This is unfortunately one of the limitations of the current system.

In the future, an extension system could be made so that themes could "extend" modules in such a way that "extended" assets would only be loaded when that module is enabled. It would work like this:

![module asset extension](https://user-images.githubusercontent.com/1009343/45761284-39da7080-bc2c-11e8-8eba-ca94a8c5dd45.jpg)

So if you intend to keep overriding assets, you are free to continue doing it... at your own risk 😉

**Comment by dennispw** [comment]
I see what you're saying but I also don't see what you're saying. :)

I understand that an override essentially is a complete replacement but I would still argue that it's a better way to go since you never touch the actual module. Sure, any designer/developer can make enough changes in which a future update might render a module broken but I assume that a person savvy enough to create overrides is also savvy enough to recognise when that's the case and make the proper adjustments. After all that's not unique to the module overrides but also a theme. A theme made for a version of Prestashop might break with the next hence the makers have to be on top of this. It comes with the territory so to speak.

But anyway.

At the very least the Classic theme should be slightly refactored (in my opinion) to remove certain styles and scripts wherever possible to the correct module override. Such as the `.featured-products` selectors in my previous example.

Not only to reduce and minimize any code bloat but also to serve as a slight hint to other designers and developers to say "Hey, if you're modifying a specific module consider using it's assets!", if you know what I mean.

**Comment by dennispw** [comment]
Just look at this, this is the compiled CSS that is used both for modules such as `ps_featuredproducts` as well as the regular product list:
```
#products .product-miniature .discount-amount.discount-amount, #products .product-miniature .discount-amount.discount-percentage, #products .product-miniature .discount-percentage.discount-amount, #products .product-miniature .discount-percentage.discount-percentage, #products .product-miniature .on-sale.discount-amount, #products .product-miniature .on-sale.discount-percentage, #products .product-miniature .online-only.discount-amount, #products .product-miniature .online-only.discount-percentage, #products .product-miniature .pack.discount-amount, #products .product-miniature .pack.discount-percentage, #products .product-miniature .product-flags .new.discount-amount, #products .product-miniature .product-flags .new.discount-percentage, .featured-products .product-miniature .discount-amount.discount-amount, .featured-products .product-miniature .discount-amount.discount-percentage, .featured-products .product-miniature .discount-percentage.discount-amount, .featured-products .product-miniature .discount-percentage.discount-percentage, .featured-products .product-miniature .on-sale.discount-amount, .featured-products .product-miniature .on-sale.discount-percentage, .featured-products .product-miniature .online-only.discount-amount, .featured-products .product-miniature .online-only.discount-percentage, .featured-products .product-miniature .pack.discount-amount, .featured-products .product-miniature .pack.discount-percentage, .featured-products .product-miniature .product-flags .new.discount-amount, .featured-products .product-miniature .product-flags .new.discount-percentage, .product-accessories .product-miniature .discount-amount.discount-amount, .product-accessories .product-miniature .discount-amount.discount-percentage, .product-accessories .product-miniature .discount-percentage.discount-amount, .product-accessories .product-miniature .discount-percentage.discount-percentage, .product-accessories .product-miniature .on-sale.discount-amount, .product-accessories .product-miniature .on-sale.discount-percentage, .product-accessories .product-miniature .online-only.discount-amount, .product-accessories .product-miniature .online-only.discount-percentage, .product-accessories .product-miniature .pack.discount-amount, .product-accessories .product-miniature .pack.discount-percentage, .product-accessories .product-miniature .product-flags .new.discount-amount, .product-accessories .product-miniature .product-flags .new.discount-percentage, .product-miniature .product-miniature .discount-amount.discount-amount, .product-miniature .product-miniature .discount-amount.discount-percentage, .product-miniature .product-miniature .discount-percentage.discount-amount, .product-miniature .product-miniature .discount-percentage.discount-percentage, .product-miniature .product-miniature .on-sale.discount-amount, .product-miniature .product-miniature .on-sale.discount-percentage, .product-miniature .product-miniature .online-only.discount-amount, .product-miniature .product-miniature .online-only.discount-percentage, .product-miniature .product-miniature .pack.discount-amount, .product-miniature .product-miniature .pack.discount-percentage, .product-miniature .product-miniature .product-flags .new.discount-amount, .product-miniature .product-miniature .product-flags .new.discount-percentage {
    z-index: 2;
    background: #f39d72;
}
```

```
#products .product-miniature .discount-amount, #products .product-miniature .discount-percentage, #products .product-miniature .on-sale, #products .product-miniature .online-only, #products .product-miniature .pack, #products .product-miniature .product-flags .new, .featured-products .product-miniature .discount-amount, .featured-products .product-miniature .discount-percentage, .featured-products .product-miniature .on-sale, .featured-products .product-miniature .online-only, .featured-products .product-miniature .pack, .featured-products .product-miniature .product-flags .new, .product-accessories .product-miniature .discount-amount, .product-accessories .product-miniature .discount-percentage, .product-accessories .product-miniature .on-sale, .product-accessories .product-miniature .online-only, .product-accessories .product-miniature .pack, .product-accessories .product-miniature .product-flags .new, .product-miniature .product-miniature .discount-amount, .product-miniature .product-miniature .discount-percentage, .product-miniature .product-miniature .on-sale, .product-miniature .product-miniature .online-only, .product-miniature .product-miniature .pack, .product-miniature .product-miniature .product-flags .new {
    display: block;
    position: absolute;
    left: -.4375rem;
    padding: .3125rem .4375rem;
    color: #fff;
    background: #2fb5d2;
    text-transform: uppercase;
    min-width: 3.125rem;
    min-height: 1.875rem;
    font-size: 1rem;
    font-weight: 600;
    box-shadow: 2px 2px 4px 0 rgba(0,0,0,.2);
}
```

The selectors when using these methods is *really* out of hand. Either refractor the classic SCSS to use much, much simpler selectors or lift them out to the seperate modules so unused CSS isn't loaded. (Such as disabled modules)

**Comment by matks** [comment]
ping @NeOMakinG I think you would like this topic 😉

**Comment by NeOMakinG** [comment]
As discussed with @PierreRambaud :

We think that the classic webpack configuration is managing theme assets, and should not manage modules assets which are really out of scope. 

It looks like we doesn't plane to to anything for this issue for the moment, let's close it !

Thanks for your feedback anyway, it will help us to decide on a lot of points in the next major!
